### ISSUE TITLE
Bad assumption in roledb leads to persistent freeze attack

### ISSUE THREAD

**Issue by awwad** [theupdateframework/tuf#736]
## Immediate Issue Summary
If the client concludes that its snapshot version is out of date (i.e. it has validated a new timestamp version that lists a snapshot version newer than the one it currently has and trusts), it tries to fetch the new snapshot version. If that fails (can't get snapshot or snapshot can't be validated), then the previously trusted known-to-be-outdated snapshot metadata is deleted, both on disk and in memory. This is correct behavior per the spec. Unfortunately, the in-memory deletion of the snapshot metadata stored in roledb results in an error on any further attempts to update snapshot (see below).

Credit goes to Allen Cain (@eacain) for finding this issue.

## Diagnosis
There is a bug in TUF in the functioning of tuf.sig.verify / tuf.roledb. In this bug, the keys expected to sign for role X (snapshot in this case) are stored in the roledb entry for role X. (In this case, tuf.sig.verify expects to find the keys expected of valid snapshot metadata in roledb's entry for snapshot, not root.) While this information is still updated based only on validated root metadata, this implementation remains wrong for a few reasons. The issue here is that role X may be deleted if a new version fails to validate and it is known that the current version is outdated (as correctly occurs in this case); while that's fine, the implementation cannot continue to expect to find the keys in the deleted entry. These keys come from the delegating role (root in this case), and they should be checked against information indexed by the delegating role, not indexed against the delegated role. This behavior is also related to a known issue (#660) that can cause difficulties if multiple roles delegate to the same delegated Targets role. Other solutions to the issue identified here include changing the role removal code to leave the expected keys cached there from the delegating role; this, however is not advisable as it is somewhat counter-intuitive and might lead to flawed implementation or odd edge cases (similar to this one).

In short, the information should be indexed by delegating role, not by delegated role, solving both of these issues. This issue derives from issue #660, so see that for the broader explanation of the more general issue.

**Comment by vladimir-v-diaz** [comment]
Are you able to reproduce the persistent freeze attack in the latest code?  We should have an integration test for this attack, so I'd be curious to know why it's not encountering this issue.

After reviewing the code, it appears that the role db is rebuilt (including the keys for Snapshot, Timestamp, Targets, and Root) when it validates and installs new Root metadata.  The relevant lines of code:

https://github.com/theupdateframework/tuf/blob/76ca2e83643fa01cf713e194553adb8a38113a83/tuf/client/updater.py#L1776-L1780


The persistent freeze issue should be avoided by the current code as follows:

(1) attempt to refresh top-level metadata.
(2) root indicates new versions of Timestamp and Snapshot.
(3) client is given good Timestamp metadata and installs it.
(4) client is given bad Snapshot metadata!
(5) client discards the bad Snapshot that it was given, and the stale version locally installed.
(6) client attempts to recover by checking for new metadata.
(7) a new root is given, which causes the role db to be rebuilt (i.e., Snapshot info should be up to date).
(8) download new Timestamp and Snapshot.

Note: The code attempts to update metadata if it is also missing. See https://github.com/theupdateframework/tuf/blob/76ca2e83643fa01cf713e194553adb8a38113a83/tuf/client/updater.py#L1931-L1990

I do agree that issue #660 is a problem, which will addressed in a future refactor of the low level API.

**Comment by awwad** [comment]
I was able to reproduce it in the current code, yes. I'll show you the
script.

**Comment by awwad** [comment]
Vlad and I chatted. A reasonable short-term fix to this would be ensuring that every top-level metadata refresh rebuilds roledb (even if root hasn't changed) which is easy to do.

(As a sidenote, issuing new metadata would resolve this problem if a client encountered it, for the current branch of TUF (though seemingly not the Uptane fork), limiting impact slightly.)

I'll make the short-term fix for now.

**Comment by trishankatdatadog** [comment]
Great job finding this @eacain!
